app/assets/javascripts/models/beta/story.js

```diff
@@ -298,14 +319,40 @@ export const cloneStory = (story) => {
     tasks: [],
     notes: [],
     documents: [],
-    _editing: null
-  }
+    _editing: null,
+  };
 
   return {
     ...clonedStory,
-    _editing: clonedStory
-  }
-}
+    _editing: clonedStory,
+  };
+};
+
+export const manageUserLocalEditsAndAdditions = (
+  stateStories,
+  serverStories
+) => {
+  const storyInCreationProcess = stateStories.filter(
+    (story) => story.id === null
+  );
+  const editedStories = serverStories.map((receivedStory) => {
+    const existingStory = stateStories.find(
+      (story) => story.id === receivedStory.id
+    );
+    if (existingStory && existingStory.collapsed === false) {
+      return {
+        ...receivedStory,
+        collapsed: false,
+        _editing: existingStory._editing,
+      };
+    }
+    return {
+      ...receivedStory,
+    };
+  });
+
+  return [...storyInCreationProcess, ...editedStories];
+};
```

Comment:
I really like this function! Well done. I'm gonna suggest some things:

- improving some names
- use the function `isNew` that is in this file to check if the story was not saved yet
- we don't need to spread `receivedStory` (actually, if we do, we can cause an unwanted updated on the component)

```suggestion
export const mergeWithFetchedStories = (
  currentStories,
  fetchedStories
) => {
  const storyNotSaved = currentStories.filter(isNew);
  
  const editedStories = fetchedStories.map((fetchedStory) => {
    const existingStory = currentStories.find(
      (story) => story.id === fetchedStory.id
    );
    
    if (existingStory && !existingStory.collapsed) {
      return {
        ...fetchedStory,
        collapsed: false,
        _editing: existingStory._editing,
      };
    }
    
    return fetchedStory;
  });

  return [...storyNotSaved, ...editedStories];
};
```